File: sdk/typescript/src/prompt.ts
export interface PromptArg {
  name: string;
  description?: string;
  required?: boolean;
}

export interface PromptMessage {
  role: 'user' | 'assistant';
  content: string;
}

export interface PromptDef {
  name: string;
  description: string;
  arguments: PromptArg[];
  handler: (args: Record<string, string>) => PromptMessage | PromptMessage[];
}

const promptRegistry: PromptDef[] = [];

export interface PromptOptions {
  description: string;
  name?: string;
  arguments?: PromptArg[];
  handler: (args: Record<string, string>) => PromptMessage | PromptMessage[];
}

export function prompt(options: PromptOptions): PromptDef {
  const def: PromptDef = {
    name: options.name ?? `prompt_${promptRegistry.length}`,
    description: options.description,
    arguments: options.arguments ?? [],
    handler: options.handler,
  };
  promptRegistry.push(def);
  return def;
}

export function getRegisteredPrompts(): PromptDef[] {
  return [...promptRegistry];
}

File: sdk/rust/src/prompt.rs
use std::sync::Mutex;

pub struct PromptArg {
    pub name: String,
    pub description: String,
    pub required: bool,
}

pub struct PromptMessage {
    pub role: String,
    pub content: String,
}

pub struct PromptDef {
    pub name: String,
    pub description: String,
    pub arguments: Vec<PromptArg>,
    pub handler: Box<dyn Fn(serde_json::Value) -> Vec<PromptMessage> + Send + Sync>,
}

static PROMPT_REGISTRY: Mutex<Vec<PromptDef>> = Mutex::new(Vec::new());

pub fn register_prompt(def: PromptDef) {
    PROMPT_REGISTRY.lock().unwrap_or_else(|e| e.into_inner()).push(def);
}

pub(crate) fn with_prompts<F, R>(f: F) -> R
where F: FnOnce(&[PromptDef]) -> R {
    let guard = PROMPT_REGISTRY.lock().unwrap_or_else(|e| e.into_inner());
    f(&guard)
}

pub fn clear_prompt_registry() {
    PROMPT_REGISTRY.lock().unwrap_or_else(|e| e.into_inner()).clear();
}
